manager/logbroker/matcher.go

```diff
@@ -0,0 +1,132 @@
+package logbroker
+
+import (
+	"context"
+	"sync"
+
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+	"github.com/docker/swarmkit/manager/state"
+	"github.com/docker/swarmkit/manager/state/store"
+	"github.com/docker/swarmkit/watch"
+)
+
+type newMatch struct {
+	nodeID       string
+	subscription *api.SubscriptionMessage
+}
+
+type matcher struct {
+	mu sync.RWMutex
+
+	store       *store.MemoryStore
+	sub         *api.SubscriptionMessage
+	changeQueue *watch.Queue
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	nodes map[string]struct{}
+}
+
+func newMatcher(store *store.MemoryStore, sub *api.SubscriptionMessage, changeQueue *watch.Queue) *matcher {
+	return &matcher{
+		store:       store,
+		sub:         sub,
+		changeQueue: changeQueue,
+		nodes:       make(map[string]struct{}),
+	}
+}
+
+func (m *matcher) Contains(nodeID string) bool {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	_, ok := m.nodes[nodeID]
+	return ok
+}
+
+func (m *matcher) Run(ctx context.Context) {
+	m.ctx, m.cancel = context.WithCancel(ctx)
+	m.match()
+	go m.watch()
+}
+
+func (m *matcher) Stop() {
+	if m.cancel != nil {
+		m.cancel()
+	}
+}
+
+func (m *matcher) match() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.store.View(func(tx store.ReadTx) {
+		for _, nid := range m.sub.Selector.NodeIDs {
+			m.nodes[nid] = struct{}{}
+		}
+
+		for _, tid := range m.sub.Selector.TaskIDs {
+			if task := store.GetTask(tx, tid); task != nil {
+				m.nodes[task.NodeID] = struct{}{}
+			}
+		}
+
+		for _, sid := range m.sub.Selector.ServiceIDs {
+			tasks, err := store.FindTasks(tx, store.ByServiceID(sid))
+			if err != nil {
+				log.L.Warning(err)
+				continue
+			}
+			for _, task := range tasks {
+				m.nodes[task.NodeID] = struct{}{}
+			}
+		}
+	})
+}
+
+func (m *matcher) watch() error {
+	wq := m.store.WatchQueue()
+
+	matchTasks := map[string]struct{}{}
+	for _, tid := range m.sub.Selector.TaskIDs {
+		matchTasks[tid] = struct{}{}
+	}
+
+	matchServices := map[string]struct{}{}
+	for _, sid := range m.sub.Selector.ServiceIDs {
+		matchServices[sid] = struct{}{}
+	}
+
+	ch, cancel := state.Watch(wq, state.EventUpdateTask{})
+	defer cancel()
+
+	add := func(nodeID string) {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+
+		if _, ok := m.nodes[nodeID]; !ok {
+			m.nodes[nodeID] = struct{}{}
+			m.changeQueue.Publish(&newMatch{
+				subscription: m.sub,
+				nodeID:       nodeID,
+			})
+		}
+	}
+
+	for {
+		select {
+		case <-m.ctx.Done():
+			return m.ctx.Err()
+		case e := <-ch:
+			t := e.(state.EventUpdateTask).Task
+			if _, ok := matchTasks[t.ID]; ok {
+				add(t.NodeID)
+			}
+			if _, ok := matchServices[t.ServiceID]; ok {
+				add(t.NodeID)
+			}
```

Comment:
It wouldn't be much of a problem, right? Closing subscriptions as they are not active anymore may be more expensive than just leaving them alone ... maybe?